### PR TITLE
sql: extend resolverQueue to support waiter transactions

### PR DESCRIPTION
**Pull request by Azhng**
Reviewer note: only last two commits are relevant

---

# First commit

sql: add waiter txn id to extended contention event

Previously, contention event only include the txnID of the transaction
that held the lock (a.k.a. blocking transaction). The txnID of the
waiter transaction is missing from the contention event.

This commit includes the txnID of the waiter transaction into contention
events.

Release note: None

---

# Second commit

sql: extend resolverQueue to resolve waiter txn id

Previously, the resovlerQueue used in the contention event store only
resolved the txnID of the blocking transaction.
This commit, the resolverQueue would also resolve the txnID of the
waiting transaction.

Release note: None

### CODE REVIEW

**Review by matthewtodd — CHANGES_REQUESTED** [review]
Nice! I like where this is going.

> Reviewed 2 of 5 files at r3, 1 of 7 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-:-Mw7XbFf565xwnArmFo1:b-m824mm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng and @matthewtodd)

___
*[pkg/sql/contention/resolver.go, line 107 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-Mw7RKa93GatDs4E0Ytz:-Mw7RKa93GatDs4E0Yu-:b-cy3neq) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cf0c755d2866f415de1525d5d4a2da3e92289f5/pkg/sql/contention/resolver.go#L107)):*
> ```Go
> 		// remainingRetries stores a mapping of each contention event to its
> 		// remaining number of retries attempts. The key in the map is the
> 		// transaction ID of the blocking transaction.
> ```

this comment is no longer true
___
*[pkg/sql/contention/resolver.go, line 172 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-Mw7SOHSDUjCWaYFS_Nl:-Mw7SOHSDUjCWaYFS_Nm:binjej5) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cf0c755d2866f415de1525d5d4a2da3e92289f5/pkg/sql/contention/resolver.go#L172)):*
> ```Go
> 		//  self-throttling once that QPS value exceed certain value.
> 
> 		remoteReq, localReq := makeRPCRequestsFromBatch(currentBatch)
> ```

Now I understand that the remoteReq is to resolve blocking txn ids, and the localReq is for waiting txn ids. But it took me a while. Maybe consider renaming these variables?
___
*[pkg/sql/contention/resolver.go, line 188 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-Mw7Ssj58leDGLDAxPPY:-Mw7Ssj58leDGLDAxPPZ:b-czbhkh) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cf0c755d2866f415de1525d5d4a2da3e92289f5/pkg/sql/contention/resolver.go#L188)):*
> ```Go
> 
> 		for _, event := range currentBatch {
> 			needToRetryDueToBlockingTxnID, initialRetryBudget1 :=
> ```

I wonder if a name rather than a numeric suffix would help? Could get too long, though.
___
*[pkg/sql/contention/resolver.go, line 189 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-Mw7Sr6a--rxPiUnKK5a:-Mw7Sr6a--rxPiUnKK5b:b-avw5u8) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cf0c755d2866f415de1525d5d4a2da3e92289f5/pkg/sql/contention/resolver.go#L189)):*
> ```Go
> 		for _, event := range currentBatch {
> 			needToRetryDueToBlockingTxnID, initialRetryBudget1 :=
> 				maybeUpdateBlockingTxnFingerprintID(
> ```

drop the word blocking
___
*[pkg/sql/contention/resolver.go, line 233 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-Mw7TQLK2RyqNAWc6jxh:-Mw7TQLK2RyqNAWc6jxi:b-5bplef) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cf0c755d2866f415de1525d5d4a2da3e92289f5/pkg/sql/contention/resolver.go#L233)):*
> ```Go
> 	}
> 
> 	if uuid.Nil.Equal(txnID) {
> ```

comment here?
___
*[pkg/sql/contention/resolver.go, line 242 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-Mw7ULBz44-VA5Ci2zDe:-Mw7ULBz44-VA5Ci2zDf:bvttr0m) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cf0c755d2866f415de1525d5d4a2da3e92289f5/pkg/sql/contention/resolver.go#L242)):*
> ```Go
> 
> 	if _, ok := inProgressTxnIDs[txnID]; ok {
> 		return true /* needToRetry */, retryBudgetInfinite
> ```

nit: retryBudgetForTxnInProgress
___
*[pkg/sql/contention/resolver.go, line 335 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-Mw7VfJF2CzPZu1KIJX7:-Mw7VfJF2CzPZu1KIJX8:b-yc8hzo) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cf0c755d2866f415de1525d5d4a2da3e92289f5/pkg/sql/contention/resolver.go#L335)):*
> ```Go
> func makeRPCRequestsFromBatch(
> 	batch []contentionpb.ExtendedContentionEvent,
> ) (remoteReq, localReq *serverpb.TxnIDResolutionRequest) {
> ```

consider the same renaming here
___
*[pkg/sql/contentionpb/contention.go, line 90 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-Mw7WgRT5hw96eOBm03z:-Mw7WgRT5hw96eOBm04-:bbsf82t) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cf0c755d2866f415de1525d5d4a2da3e92289f5/pkg/sql/contentionpb/contention.go#L90)):*
> ```Go
> }
> 
> func hashUUID(u uuid.UUID, fnv *util.FNV64) {
> ```

nit: since this code is a little clever, it's probably worth a direct unit test that it works? (You can craft UUID bytes to show things like first half same, second half different -> different hash, etc.)
___
*[pkg/sql/contentionpb/contention.go, line 92 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-Mw7WIQ06mWAzei55n-j:-Mw7WIQ06mWAzei55n-k:b-c80uhi) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cf0c755d2866f415de1525d5d4a2da3e92289f5/pkg/sql/contentionpb/contention.go#L92)):*
> ```Go
> func hashUUID(u uuid.UUID, fnv *util.FNV64) {
> 	b := u.GetBytes()
> 	// Hash the blocking txn id.
> ```

maybe say more about what's happening here, why 2 calls to DecodeUint64Descending
___
*[pkg/sql/contention/resolver_test.go, line 415 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-Mw7XIAIE3F-uJ7aE1xx:-Mw7XIAIE3F-uJ7aE1xy:b-qly58a) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cf0c755d2866f415de1525d5d4a2da3e92289f5/pkg/sql/contention/resolver_test.go#L415)):*
> ```Go
> 		f.setTxnIDEntry(tc.coordinatorNodeID, tc.TxnID, tc.TxnFingerprintID)
> 		// TODO(azhng): wip: this is kind of a hack heh?
> 		f.setTxnIDEntry("local", tc.TxnID, tc.TxnFingerprintID)
> ```

delete this? Also add some more test cases: want to populate remote & local coordinators with different data? Also test the retry behavior.

**Review by Azhng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76719)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @matthewtodd)

___
*[pkg/sql/contention/resolver.go, line 107 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-Mw7RKa93GatDs4E0Ytz:-Mw881oQDKQqJGbKevTV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cf0c755d2866f415de1525d5d4a2da3e92289f5/pkg/sql/contention/resolver.go#L107)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

this comment is no longer true
</blockquote></details>

Done.
___
*[pkg/sql/contention/resolver.go, line 172 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-Mw7SOHSDUjCWaYFS_Nl:-Mw88j497OamWdWoejIH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cf0c755d2866f415de1525d5d4a2da3e92289f5/pkg/sql/contention/resolver.go#L172)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

Now I understand that the remoteReq is to resolve blocking txn ids, and the localReq is for waiting txn ids. But it took me a while. Maybe consider renaming these variables?
</blockquote></details>

Done.
___
*[pkg/sql/contention/resolver.go, line 188 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-Mw7Ssj58leDGLDAxPPY:-Mw88kUH7-68JsjgEJpd:b-xgqf3h) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cf0c755d2866f415de1525d5d4a2da3e92289f5/pkg/sql/contention/resolver.go#L188)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

I wonder if a name rather than a numeric suffix would help? Could get too long, though.
</blockquote></details>

Done. renamed 
___
*[pkg/sql/contention/resolver.go, line 189 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-Mw7Sr6a--rxPiUnKK5a:-Mw870Zh0vm6j7oVHAj1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cf0c755d2866f415de1525d5d4a2da3e92289f5/pkg/sql/contention/resolver.go#L189)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

drop the word blocking
</blockquote></details>

Done.
___
*[pkg/sql/contention/resolver.go, line 233 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-Mw7TQLK2RyqNAWc6jxh:-Mw87g_Z5p8-PBHw6LMm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cf0c755d2866f415de1525d5d4a2da3e92289f5/pkg/sql/contention/resolver.go#L233)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

comment here?
</blockquote></details>

Done.
___
*[pkg/sql/contention/resolver.go, line 242 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-Mw7ULBz44-VA5Ci2zDe:-Mw87oeP7IP5MjlWldJH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cf0c755d2866f415de1525d5d4a2da3e92289f5/pkg/sql/contention/resolver.go#L242)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

nit: retryBudgetForTxnInProgress
</blockquote></details>

Done.
___
*[pkg/sql/contention/resolver.go, line 335 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-Mw7VfJF2CzPZu1KIJX7:-Mw88vYwAa6iR5XxBi4l:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cf0c755d2866f415de1525d5d4a2da3e92289f5/pkg/sql/contention/resolver.go#L335)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

consider the same renaming here
</blockquote></details>

Done.
___
*[pkg/sql/contentionpb/contention.go, line 90 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-Mw7WgRT5hw96eOBm03z:-Mw8BiX44owcOGELoTe_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cf0c755d2866f415de1525d5d4a2da3e92289f5/pkg/sql/contentionpb/contention.go#L90)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

nit: since this code is a little clever, it's probably worth a direct unit test that it works? (You can craft UUID bytes to show things like first half same, second half different -> different hash, etc.)
</blockquote></details>

Done.
___
*[pkg/sql/contentionpb/contention.go, line 92 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-Mw7WIQ06mWAzei55n-j:-Mw8BjCoCWYsd8a-dQy0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cf0c755d2866f415de1525d5d4a2da3e92289f5/pkg/sql/contentionpb/contention.go#L92)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

maybe say more about what's happening here, why 2 calls to DecodeUint64Descending
</blockquote></details>

Done.
___
*[pkg/sql/contention/resolver_test.go, line 415 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-Mw7XIAIE3F-uJ7aE1xx:-Mw8rT1d1AGMpFJfpa7T:buymgh3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cf0c755d2866f415de1525d5d4a2da3e92289f5/pkg/sql/contention/resolver_test.go#L415)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

delete this? Also add some more test cases: want to populate remote & local coordinators with different data? Also test the retry behavior.
</blockquote></details>

Done. Extended test harness to support have different waiting/blocking txn metadata.

File: pkg/sql/contention/event_store.go [hunk 1]
```diff
@@ -274,12 +272,12 @@ func (s *eventStore) forEachEvent(
 }
 
 func (s *eventStore) getEventByBlockingTxnID(
-	txnID uuid.UUID,
+	hash uint64,
```
Comment (THardy98) [inline pkg/sql/contention/event_store.go:275]: This function needs renaming (by Blocking + Waiting + Timestamp hash ID)

**Review by THardy98 — CHANGES_REQUESTED** [review]

File: pkg/sql/contention/event_store.go (hunk 1 above)
Comment (Azhng) [reply pkg/sql/contention/event_store.go:275]: Renamed.

**Review by matthewtodd — APPROVED** [review]
Rebasing these stacked PRs makes finding just the new code confusing -- maybe something to strategize around in the future? But I think I found it here! Just a few nits, but otherwise :lgtm:

> Reviewed 1 of 5 files at r10, 5 of 8 files at r11, 1 of 5 files at r14, 1 of 8 files at r15.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-:-MwhTZYA4mZl6-X2iEsB:b9pdjq2)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @Azhng, @matthewtodd, and @THardy98)

___
*[pkg/sql/contention/resolver.go, line 172 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-MwhTKA7430ZEvhOwdP0:-MwhTKA86DVYwWYXqD0H:bhpsrdr) ([raw file](https://github.com/cockroachdb/cockroach/blob/c82d6a7a34aeea74f67e1007698823da0afad0d4/pkg/sql/contention/resolver.go#L172)):*
> ```Go
> 		//  self-throttling once that QPS value exceed certain value.
> 
> 		blockTxnIDsRPCReq, waitingTxnIDsRPCReq := makeRPCRequestsFromBatch(currentBatch)
> ```

nit: maybe block -> blocking
___
*[pkg/sql/contention/resolver.go, line 172 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-MwhTM_Q6PKrJoyccl20:-MwhTM_Q6PKrJoyccl21:b-2blzh) ([raw file](https://github.com/cockroachdb/cockroach/blob/c82d6a7a34aeea74f67e1007698823da0afad0d4/pkg/sql/contention/resolver.go#L172)):*
> ```Go
> 		//  self-throttling once that QPS value exceed certain value.
> 
> 		blockTxnIDsRPCReq, waitingTxnIDsRPCReq := makeRPCRequestsFromBatch(currentBatch)
> ```

nit: maybe remove the `RPC` from these names? It's redundant with `Req` IMO.
___
*[pkg/sql/contention/resolver.go, line 239 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-MwhTSJVAZdL95TugG73:-MwhTSJVAZdL95TugG74:b3tc9k9) ([raw file](https://github.com/cockroachdb/cockroach/blob/c82d6a7a34aeea74f67e1007698823da0afad0d4/pkg/sql/contention/resolver.go#L239)):*
> ```Go
> 	}
> 
> 	// Sometimes DistSQL engine is used by in weird ways. It is possible for a
> ```

nit: "is used by in"
___
*[pkg/sql/contention/resolver.go, line 343 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-MwhTVO_4HoyJWFqrqX6:-MwhTVO_4HoyJWFqrqX7:b1d0m7v) ([raw file](https://github.com/cockroachdb/cockroach/blob/c82d6a7a34aeea74f67e1007698823da0afad0d4/pkg/sql/contention/resolver.go#L343)):*
> ```Go
> func makeRPCRequestsFromBatch(
> 	batch []contentionpb.ExtendedContentionEvent,
> ) (blockingTxnIDRPCReq, waitingTxnIDRPCReq *serverpb.TxnIDResolutionRequest) {
> ```

nit: ditto re: the redundancy of `RPC` here.

**Review by Azhng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76719)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @matthewtodd and @THardy98)

___
*[pkg/sql/contention/resolver.go, line 172 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-MwhTKA7430ZEvhOwdP0:-MwhVOqT50lezvSTp5tb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c82d6a7a34aeea74f67e1007698823da0afad0d4/pkg/sql/contention/resolver.go#L172)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

nit: maybe block -> blocking
</blockquote></details>

Done.
___
*[pkg/sql/contention/resolver.go, line 172 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-MwhTM_Q6PKrJoyccl20:-MwhVfaPEdq_AnEws620:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c82d6a7a34aeea74f67e1007698823da0afad0d4/pkg/sql/contention/resolver.go#L172)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

nit: maybe remove the `RPC` from these names? It's redundant with `Req` IMO.
</blockquote></details>

Done.
___
*[pkg/sql/contention/resolver.go, line 239 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-MwhTSJVAZdL95TugG73:-MwhVkYX5keyQGMs12O9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c82d6a7a34aeea74f67e1007698823da0afad0d4/pkg/sql/contention/resolver.go#L239)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

nit: "is used by in"
</blockquote></details>

Done.
___
*[pkg/sql/contention/resolver.go, line 343 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/76719#-MwhTVO_4HoyJWFqrqX6:-MwhVluKBafepv0hbpe1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c82d6a7a34aeea74f67e1007698823da0afad0d4/pkg/sql/contention/resolver.go#L343)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

nit: ditto re: the redundancy of `RPC` here.
</blockquote></details>

Done.

**Comment by Azhng** [comment]
TFTR! 

bors r+

**Pull request merged**